File: OOPL/lab2.java
import java.util.Scanner;

class lab2{
    public static Scanner sc = new Scanner(System.in);
    public static int[] createArray(){
        System.out.print("Enter number of elements in array: ");
        int n = sc.nextInt();
        int array[] = new int[n];
        for(int i=0;i<n;i++){
            System.out.print("Enter element at index "+i+" :");
            array[i] = sc.nextInt();
        }

        return array;
    }

    public static int[] bubbleSort(int[] array,boolean reverse){
        int n = array.length;
        if(!reverse){
            for(int i=0;i<n-1;i++){
                for(int j=0;j<n-i-1;j++){
                    if(array[j]>array[j+1]){
                        int temp = array[j];
                        array[j] = array[j+1];
                        array[j+1] = temp;
                    }
                }
            }
        }
        if(reverse){
            for(int i=0;i<n-1;i++){
                for(int j=0;j<n-i-1;j++){
                    if(array[j]<array[j+1]){
                        int temp = array[j];
                        array[j] = array[j+1];
                        array[j+1] = temp;
                    }
                }
            }
        }
        System.out.println("The array in "+ (reverse?"descending":"ascending")+" order is:");
        for(int i=0;i<n;i++){
            System.out.print(array[i]+" ");
        }

        return array;
    }

    public static int[][] createMatrix(int n,int m){
        int matrix[][] = new int[n][m];
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                System.out.print("Enter element at row "+i+" column "+j+": ");
                matrix[i][j] = sc.nextInt();
            }
        }
        System.out.println("The matrix is:");
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                System.out.print(matrix[i][j]+"\t");
            }
            System.out.print("\n");
        }

        return matrix;
    }

	public static void main(String args[]){
        while(true){
            System.out.print("Enter question no.: ");
            int q = sc.nextInt();
            
            if(q==1){
                int arr[] = createArray();
                System.out.print("Ascending or descending? a/d: ");
                char order = sc.next().charAt(0);
                if(order=='a'){
                    bubbleSort(arr, false);
                }
                else if(order=='d' || order=='D'){
                    bubbleSort(arr, true);
                }
            }
            else if(q==2){
                System.out.print("Enter number of elements in array: ");
                int n = sc.nextInt();
                int arr[] = new int[n+1];
                for(int i=0;i<n;i++){
                    System.out.print("Enter element at index "+i+" :");
                    arr[i] = sc.nextInt();
                }

                System.out.print("Enter index to insert element at: ");
                int pos1 = sc.nextInt();
                System.out.print("Enter the element to be inserted: ");
                int ele = sc.nextInt();

                for(int i=n-1;i>=pos1;i--){
                    arr[i+1] = arr[i];
                }
                arr[pos1] = ele;
                System.out.println("The modified array is:");
                for(int i=0;i<arr.length;i++){
                    System.out.print(arr[i]+" ");
                }

                System.out.print("Enter position at which you want to delete element: ");
                int pos2 = sc.nextInt();
                for(int i = pos2 ; i<arr.length-1 ;i++){
                    arr[i] = arr[i+1];
                }

                System.out.println("The modified array is:");
                for(int i=0;i<arr.length-1;i++){
                    System.out.print(arr[i]+" ");
                }
            }
            else if(q==3){
                int arr[] = createArray();
                System.out.print("Enter element to be searched: ");
                int ele = sc.nextInt();
                int flag = 0;
                for(int i= 0;i<arr.length;i++){
                    if(arr[i]==ele){
                        System.out.print("The element is at index: "+i);
                        flag = 1;
                        break;
                    }
                }
                if(flag==0){System.out.println("Element does not exist in array.");}
            }
            else if(q==4){
                System.out.print("Enter number of rows of matrix: ");
                int rows1 = sc.nextInt();
                System.out.print("Enter number of columns of matrix: ");
                int columns1 = sc.nextInt();
                int mat1[][] = createMatrix(rows1, columns1);

                System.out.print("Enter number of rows of matrix: ");
                int rows2 = sc.nextInt();
                System.out.print("Enter number of columns of matrix: ");
                int columns2 = sc.nextInt();
                int mat2[][] = createMatrix(rows2, columns2);

                int newMatrix[][] = new int[rows1][columns1];
                if(rows1==rows2 && columns1==columns2){
                    for(int i=0;i<rows1;i++){
                        for(int j=0;j<columns1;j++){
                            newMatrix[i][j] = mat1[i][j] + mat2[i][j];
                        }
                    }
                    System.out.println("The matrix after addition is:");
                    for(int i=0;i<rows1;i++){
                        for(int j=0;j<columns1;j++){
                            System.out.print(newMatrix[i][j]+"\t");
                        }
                        System.out.print("\n");
                    }
                }
                else{
                    System.out.println("The dimensions of the 2 matrices are different.");
                }
            }
            else{break;}
            System.out.print("\n");
        }
        sc.close();
	}
}

File: DSAL/Lab2.c
#include <stdio.h>
#include <stdlib.h>

void Reverse(int *num,int n){
    int i,temp;
    int *first;
    int *last = (num+n-1);
    for(first=num;first<=last ; first++,last--){
        temp = *last;
        *last = *first;
        *first = temp;
    }
    for(i=0;i<n;i++){
        printf("%d",num[i]);
    }
}

int Smallest(int* array,int n){
    int smallest = *array;

    for(int i=0; i<n; i++){
        if(smallest>array[i]){
            smallest = array[i];
        }
    }

    return smallest;
}

void add(int *a,int *b){
    printf("%d",*a+*b);
}

int* biggest(int a,int b,int c){
    int bigVar = (a>b?(a>c?a:c):(b>c)?b:c);
    int *big = &bigVar;
    return big;
}

int main(void){
    int q,i,n;
    while(1){
        printf("Enter question no.: ");
        scanf("%d",&q);
        if(q==1){
            printf("Enter the length: ");
            scanf("%d",&n);
            int arr[n];
            for(int i=0;i<n;i++){
                printf("Enter element at index %d: ",i);
                scanf("%d",&arr[i]);
            }
            Reverse(arr,n);            
        }
        else if(q==2){
            printf("Enter the length: ");
            scanf("%d",&n);
            int *arr = (int*)calloc(n,sizeof(int));
            for(int i=0;i<n;i++){
                printf("Enter element at index %d: ",i);
                scanf("%d",&arr[i]);
            }
            int smallest = Smallest(arr,n);
            printf("Smallest element is %d",smallest);
        }
        else if(q==3){
            int x = 5,y=10,z=20;
            int *px = &x,*py = &y, *pz = &z;
            int **ppx = &px;
            add(px,py);
            printf("\nBiggest number is: %d",*biggest(x,y,z));
            printf("\nx = %d\npointer to x = %p\npointer to pointer to x = %p",x,px,ppx);
        }
        else if(q==4){
            int j;
            int r1,r2,c1,c2;
            printf("Enter number of rows and columns: ");
            scanf("%d %d",&r1,&c1);
            int *m1 = (int*)calloc(r1,sizeof(int));
            int *m2 = (int*)calloc(r2,sizeof(int));
            for(i=0;i<r1;i++){
                for(j=0;j<c1;j++){
                    m1 = (int*)calloc(c1,sizeof(int));
                    printf("Row %d Column %d : ",i,j);
                    scanf("%d",(m1+i)+j);
                }
            }
            for(i=0;i<r1;i++){
                for(j=0;j<c1;j++){
                    m1 = (int*)calloc(c1,sizeof(int));
                    printf("Row %d Column %d : ",i,j);
                    scanf("%d",(m1+i)+j);
                }
            }
        }
        else{
            printf("Exitting");
            break;
        }
        printf("\n");
    }

    return 0;
}